File: video_analyze/TestAnalyze.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.stream.Collectors;

/**
 * Video Analyze API 测试（Java 版）
 * 对应 video_analyze/test_analyze.py 当前版本：
 * 1) 健康检查
 * 2) 参数校验
 * 3) 同步分析
 * 4) 10 并发异步任务 + 每 5 秒轮询并打印返回参数
 */
public class TestAnalyze {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient CLIENT = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    private static final String BASE_URL = "http://10.1.6.76:9001";
    private static final String VIDEO_URL = "https://elex-apx-download.oss-cn-beijing.aliyuncs.com/material_sensortower/5a/5ae0beb8fa34103f.mp4";

    private static final int SYNC_TIMEOUT_SECONDS = 600;
    private static final int POLL_INTERVAL_SECONDS = 5;
    private static final int POLL_MAX_WAIT_SECONDS = 600;
    private static final int ASYNC_TASK_COUNT = 10;

    private static class ApiResp {
        final int httpStatus;
        final JsonNode json;
        final String raw;

        ApiResp(int httpStatus, JsonNode json, String raw) {
            this.httpStatus = httpStatus;
            this.json = json;
            this.raw = raw;
        }
    }

    private static class SubmitResult {
        final int idx;
        final boolean ok;
        final Integer httpStatus;
        final Integer code;
        final String message;
        final String taskId;

        SubmitResult(int idx, boolean ok, Integer httpStatus, Integer code, String message, String taskId) {
            this.idx = idx;
            this.ok = ok;
            this.httpStatus = httpStatus;
            this.code = code;
            this.message = message;
            this.taskId = taskId;
        }
    }

    private static void separator(String title) {
        System.out.println();
        System.out.println("============================================================");
        System.out.println("  " + title);
        System.out.println("============================================================");
    }

    private static String pretty(JsonNode node) {
        try {
            return MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(node);
        } catch (Exception e) {
            return String.valueOf(node);
        }
    }

    private static ApiResp getJson(String url, int timeoutSeconds) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(timeoutSeconds))
                .GET()
                .build();
        HttpResponse<String> resp = CLIENT.send(req, HttpResponse.BodyHandlers.ofString());
        JsonNode json = null;
        try {
            json = MAPPER.readTree(resp.body());
        } catch (Exception ignore) {
        }
        return new ApiResp(resp.statusCode(), json, resp.body());
    }

    private static ApiResp postJson(String url, JsonNode body, int timeoutSeconds) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(timeoutSeconds))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
                .build();
        HttpResponse<String> resp = CLIENT.send(req, HttpResponse.BodyHandlers.ofString());
        JsonNode json = null;
        try {
            json = MAPPER.readTree(resp.body());
        } catch (Exception ignore) {
        }
        return new ApiResp(resp.statusCode(), json, resp.body());
    }

    // 3) 异步任务并发提交 + 轮询
    private static SubmitResult submitAsyncTask(String submitUrl, JsonNode payload, int idx) {
        try {
            ApiResp resp = postJson(submitUrl, payload, 30);
            if (resp.json == null) {
                return new SubmitResult(idx, false, resp.httpStatus, null, "非 JSON 响应", null);
            }
            Integer code = resp.json.path("code").isInt() ? resp.json.path("code").asInt() : null;
            String message = resp.json.path("message").asText(null);
            String taskId = resp.json.path("data").path("task_id").asText(null);
            boolean ok = resp.httpStatus == 200 && code != null && code == 200 && taskId != null && !taskId.isBlank();
            return new SubmitResult(idx, ok, resp.httpStatus, code, message, ok ? taskId : null);
        } catch (Exception e) {
            return new SubmitResult(idx, false, null, null, e.getMessage(), null);
        }
    }

    private static boolean testAsyncTask() {
        separator("3. 异步任务并发测试  POST /api/video-analyze/tasks × " + ASYNC_TASK_COUNT);
        String submitUrl = BASE_URL + "/api/video-analyze/tasks";
        ObjectNode payload = MAPPER.createObjectNode();
        payload.put("video_url", VIDEO_URL);

        System.out.println("并发提交任务数: " + ASYNC_TASK_COUNT);
        try {
            long submitStart = System.currentTimeMillis();
            ExecutorService pool = Executors.newFixedThreadPool(ASYNC_TASK_COUNT);

                   
            List<Callable<SubmitResult>> calls = new ArrayList<>();
            for (int i = 1; i <= ASYNC_TASK_COUNT; i++) {
                final int idx = i;
                calls.add(() -> submitAsyncTask(submitUrl, payload, idx));
            }

            List<Future<SubmitResult>> futures = pool.invokeAll(calls);
            pool.shutdown();

            List<SubmitResult> submitResults = new ArrayList<>();
            for (Future<SubmitResult> f : futures) {
                try {
                    submitResults.add(f.get());
                } catch (ExecutionException e) {
                    submitResults.add(new SubmitResult(-1, false, null, null, e.getMessage(), null));
                }
            }
            submitResults.sort((a, b) -> Integer.compare(a.idx, b.idx));

            double submitElapsed = (System.currentTimeMillis() - submitStart) / 1000.0;
            System.out.printf("提交完成，耗时: %.2fs%n", submitElapsed);
            for (SubmitResult r : submitResults) {
                String icon = r.ok ? "✅" : "❌";
                System.out.printf("  %s #%02d http=%s code=%s task_id=%s%n",
                        icon, r.idx, String.valueOf(r.httpStatus), String.valueOf(r.code), String.valueOf(r.taskId));
            }

            List<SubmitResult> failedSubmit = submitResults.stream()
                    .filter(x -> !x.ok)
                    .collect(Collectors.toList());
            if (!failedSubmit.isEmpty()) {
                System.out.printf("%n❌ 存在提交失败任务: %d/%d%n", failedSubmit.size(), ASYNC_TASK_COUNT);
                for (SubmitResult r : failedSubmit) {
                    System.out.printf("  - #%02d: %s%n", r.idx, r.message);
                }
                return false;
            }

            List<String> taskIds = submitResults.stream()
                    .map(r -> r.taskId)
                    .collect(Collectors.toList());
            System.out.printf("%n开始轮询 %d 个任务（间隔 %ds，最多等待 %ds）...%n%n",
                    taskIds.size(), POLL_INTERVAL_SECONDS, POLL_MAX_WAIT_SECONDS);

            Map<String, String> taskStatus = new LinkedHashMap<>();
            Map<String, JsonNode> taskResult = new LinkedHashMap<>();
            for (String tid : taskIds) {
                taskStatus.put(tid, "pending");
            }

            long pollStart = System.currentTimeMillis();
            while ((System.currentTimeMillis() - pollStart) / 1000.0 < POLL_MAX_WAIT_SECONDS) {
                long cycleStart = System.currentTimeMillis();
                double elapsed = (System.currentTimeMillis() - pollStart) / 1000.0;
                int done = 0;
                int failed = 0;

                for (String tid : taskIds) {
                    String current = taskStatus.get(tid);
                    if ("completed".equals(current) || "failed".equals(current)) {
                        if ("completed".equals(current)) {
                            done++;
                        } else {
                            failed++;
                        }
                        continue;
                    }

                    ApiResp resp = getJson(BASE_URL + "/api/video-analyze/tasks/" + tid, 10);
                    if (resp.json == null) {
                        System.out.printf("    返回参数 task_id=%s code=null message=非JSON status=unknown error_code=null error=%s duration_seconds=null%n",
                                tid, resp.raw);
                        continue;
                    }

                    JsonNode taskData = resp.json.path("data");
                    String status = taskData.path("status").asText("unknown");
                    taskStatus.put(tid, status);

                    System.out.printf(
                            "    返回参数 task_id=%s code=%s message=%s status=%s error_code=%s error=%s duration_seconds=%s%n",
                            tid,
                            safeText(resp.json.get("code")),
                            safeText(resp.json.get("message")),
                            safeText(taskData.get("status")),
                            safeText(taskData.get("error_code")),
                            safeText(taskData.get("error")),
                            safeText(taskData.get("duration_seconds"))
                    );

                    if ("completed".equals(status) || "failed".equals(status)) {
                        taskResult.put(tid, taskData);
                    }
                    if ("completed".equals(status)) {
                        done++;
                    } else if ("failed".equals(status)) {
                        failed++;
                    }
                }

                System.out.printf("  [%.1fs] completed=%d/%d failed=%d/%d%n",
                        elapsed, done, taskIds.size(), failed, taskIds.size());

                if (done + failed == taskIds.size()) {
                    System.out.println();
                    System.out.println("全部任务进入最终态。");
                    System.out.println("  完成: " + done);
                    System.out.println("  失败: " + failed);

                    if (failed > 0) {
                        System.out.println("❌ 异步任务并发测试失败（存在 failed）");
                        for (Map.Entry<String, JsonNode> e : taskResult.entrySet()) {
                            JsonNode data = e.getValue();
                            if ("failed".equals(data.path("status").asText())) {
                                System.out.printf("  - %s: %s%n", e.getKey(), data.path("error").asText(""));
                            }
                        }
                        return false;
                    }

                    String firstTid = taskIds.get(0);
                    JsonNode sample = taskResult.get(firstTid);
                    System.out.printf("%n样例任务结果（task_id=%s）:%n", firstTid);
                    System.out.println(pretty(sample.path("result")));
                    System.out.println("✅ 异步任务并发测试成功");
                    return true;
                }

                long spentMs = System.currentTimeMillis() - cycleStart;
                long sleepMs = Math.max(0, POLL_INTERVAL_SECONDS * 1000L - spentMs);
                Thread.sleep(sleepMs);
            }

            List<String> unfinished = taskIds.stream()
                    .filter(tid -> {
                        String s = taskStatus.get(tid);
                        return !"completed".equals(s) && !"failed".equals(s);
                    })
                    .collect(Collectors.toList());
            System.out.printf("❌ 轮询超时（等待 %ds），未完成任务数: %d%n", POLL_MAX_WAIT_SECONDS, unfinished.size());
            for (String tid : unfinished) {
                System.out.printf("  - %s: %s%n", tid, taskStatus.get(tid));
            }
            return false;
        } catch (Exception e) {
            System.out.println("❌ 异步任务异常: " + e.getMessage());
            return false;
        }
    }

    // 4) 参数校验
    private static boolean testValidation() {
        separator("4. 参数校验");
        String url = BASE_URL + "/api/video-analyze/analyze";
        List<Boolean> results = new ArrayList<>();

        // 4a. 空 URL
        System.out.println("\n--- 4a. 空 video_url ---");
        try {
            ObjectNode body = MAPPER.createObjectNode();
            body.put("video_url", "");
            ApiResp resp = postJson(url, body, 10);
            System.out.println("Status: " + resp.httpStatus);
            if (resp.json != null) {
                System.out.println(pretty(resp.json));
            } else {
                System.out.println(resp.raw);
            }
            if (resp.httpStatus == 422) {
                System.out.println("✅ 空 URL 校验通过（返回 422）");
                results.add(true);
            } else {
                System.out.println("⚠ 预期 422, 实际 " + resp.httpStatus);
                results.add(false);
            }
        } catch (Exception e) {
            System.out.println("❌ 异常: " + e.getMessage());
            results.add(false);
        }

        // 4b. 非 http(s) URL
        System.out.println("\n--- 4b. 非 http(s) URL ---");
        try {
            ObjectNode body = MAPPER.createObjectNode();
            body.put("video_url", "ftp://example.com/video.mp4");
            ApiResp resp = postJson(url, body, 10);
            System.out.println("Status: " + resp.httpStatus);
            if (resp.json != null) {
                System.out.println(pretty(resp.json));
            } else {
                System.out.println(resp.raw);
            }
            if (resp.httpStatus == 422) {
                System.out.println("✅ 非 http(s) URL 校验通过（返回 422）");
                results.add(true);
            } else {
                System.out.println("⚠ 预期 422, 实际 " + resp.httpStatus);
                results.add(false);
            }
        } catch (Exception e) {
            System.out.println("❌ 异常: " + e.getMessage());
            results.add(false);
        }

        // 4c. 缺少 video_url 字段
        System.out.println("\n--- 4c. 缺少 video_url ---");
        try {
            ObjectNode body = MAPPER.createObjectNode();
            ApiResp resp = postJson(url, body, 10);
            System.out.println("Status: " + resp.httpStatus);
            if (resp.json != null) {
                System.out.println(pretty(resp.json));
            } else {
                System.out.println(resp.raw);
            }
            if (resp.httpStatus == 422) {
                System.out.println("✅ 缺少字段校验通过（返回 422）");
                results.add(true);
            } else {
                System.out.println("⚠ 预期 422, 实际 " + resp.httpStatus);
                results.add(false);
            }
        } catch (Exception e) {
            System.out.println("❌ 异常: " + e.getMessage());
            results.add(false);
        }

        return results.stream().allMatch(Boolean::booleanValue);
    }

    private static String safeText(JsonNode node) {
        if (node == null || node.isMissingNode() || node.isNull()) {
            return "null";
        }
        if (node.isTextual()) {
            return node.asText();
        }
        return node.toString();
    }

    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("  Video Analyze API 测试 (Java)");
        System.out.println("  Base URL: " + BASE_URL);
        System.out.println("  Video:    " + VIDEO_URL);
        System.out.println("============================================================");

        Map<String, Boolean> results = new LinkedHashMap<>();

        results.put("健康检查", testHealth());
        if (!results.get("健康检查")) {
            System.out.println("\n⛔ 服务不可用，跳过后续测试");
            System.exit(1);
            return;
        }

        results.put("参数校验", testValidation());
        results.put("同步分析", testSyncAnalyze());
        results.put("异步任务", testAsyncTask());

        separator("测试结果汇总");
        int passed = 0;
        for (Map.Entry<String, Boolean> e : results.entrySet()) {
            String icon = e.getValue() ? "✅" : "❌";
            System.out.printf("  %s  %s%n", icon, e.getKey());
            if (e.getValue()) {
                passed++;
            }
        }
        System.out.printf("%n  总计: %d/%d 通过%n", passed, results.size());
        System.exit(passed == results.size() ? 0 : 1);
    }
}

File: video_analyze/test_analyze.py
"""
Video Analyze Clip API 测试脚本

测试接口：
  POST /api/video-analyze/clip
  POST /api/video-analyze/clip/tasks
  GET  /api/video-analyze/clip/tasks/{task_id}
"""

import io
import json
import sys
import time
from typing import Any

# 修复 Windows GBK 控制台编码问题
if sys.stdout.encoding != "utf-8":
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8", errors="replace")
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding="utf-8", errors="replace")

import requests

BASE_URL = "http://10.1.6.76:9001"
VIDEO_URL = "https://vidio-1300638412.cos.ap-beijing.myqcloud.com/00cb5464426850d5_17.mp4"
CLIP_PROMPT = "切片尽量控制在 3 到 8 秒"

# 超时（秒）：LLM 处理视频可能耗时较长
SYNC_TIMEOUT = 600
SUBMIT_TIMEOUT = 30
POLL_TIMEOUT = 10
POLL_INTERVAL = 5
POLL_MAX_WAIT = 600
ASYNC_TASK_COUNT = 3


def separator(title: str) -> None:
    print(f"\n{'=' * 60}")
    print(f"  {title}")
    print(f"{'=' * 60}")


def pretty(data: Any) -> None:
    print(json.dumps(data, indent=2, ensure_ascii=False))


def clip_payload() -> dict[str, str]:
    return {
        "video_url": VIDEO_URL,
        "prompt": CLIP_PROMPT,
    }


def validate_clip_result(result: Any) -> tuple[bool, str]:
    if not isinstance(result, dict):
        return False, f"result 应为 object，实际为 {type(result).__name__}"

    instructions = result.get("instructions")
    if not isinstance(instructions, list) or not instructions:
        return False, "result.instructions 应为非空数组"

    emotion = result.get("emotion")
    if not isinstance(emotion, str) or not emotion.strip():
        return False, "result.emotion 应为非空字符串"

    for idx, item in enumerate(instructions):
        if not isinstance(item, dict):
            return False, f"instructions[{idx}] 应为 object"
        for field in ("start", "end", "time_str", "content"):
            if field not in item:
                return False, f"instructions[{idx}] 缺少字段 {field}"
        if not isinstance(item["start"], int) or not isinstance(item["end"], int):
            return False, f"instructions[{idx}] start/end 应为整数"
        if item["end"] <= item["start"]:
            return False, f"instructions[{idx}] 时间范围非法"
        if not str(item["time_str"]).strip() or not str(item["content"]).strip():
            return False, f"instructions[{idx}] time_str/content 不能为空"

    return True, "ok"


def test_sync_clip() -> bool:
    separator("1. 同步切片分析  POST /api/video-analyze/clip")
    url = f"{BASE_URL}/api/video-analyze/clip"
    payload = clip_payload()

    print(f"请求 URL:  {url}")
    print(f"视频地址:  {VIDEO_URL}")
    print(f"提示词:    {CLIP_PROMPT}")
    print(f"超时设置:  {SYNC_TIMEOUT}s")
    print("请求中（LLM 切片分析可能需要数分钟）...")

    start = time.time()
    try:
        resp = requests.post(url, json=payload, timeout=SYNC_TIMEOUT)
        elapsed = time.time() - start
        print(f"\nStatus: {resp.status_code}  耗时: {elapsed:.1f}s")

        data = resp.json()
        pretty(data)

        if resp.status_code != 200 or data.get("code") != 200:
            print(f"❌ 同步切片接口失败: http={resp.status_code} code={data.get('code')}")
            return False

        ok, message = validate_clip_result(data.get("data"))
        if not ok:
            print(f"❌ 同步切片返回结构错误: {message}")
            return False

        print("✅ 同步切片分析成功")
        return True

    except requests.exceptions.Timeout:
        elapsed = time.time() - start
        print(f"❌ 同步切片分析超时 ({elapsed:.1f}s)")
        return False
    except Exception as e:
        print(f"❌ 同步切片分析异常: {e}")
        return False


def test_submit_clip_tasks() -> list[str]:
    separator(f"2. 提交切片任务  POST /api/video-analyze/clip/tasks × {ASYNC_TASK_COUNT}")
    url = f"{BASE_URL}/api/video-analyze/clip/tasks"
    payload = clip_payload()
    task_ids: list[str] = []

    print(f"请求 URL: {url}")
    for idx in range(1, ASYNC_TASK_COUNT + 1):
        print(f"\n--- 提交第 {idx}/{ASYNC_TASK_COUNT} 个切片任务 ---")
        try:
            resp = requests.post(url, json=payload, timeout=SUBMIT_TIMEOUT)
            print(f"Status: {resp.status_code}")

            data = resp.json()
            pretty(data)

            if resp.status_code != 200 or data.get("code") != 200:
                print(f"❌ 提交切片任务失败: http={resp.status_code} code={data.get('code')}")
                return task_ids

            task_id = ((data.get("data") or {}).get("task_id") or "").strip()
            status = (data.get("data") or {}).get("status")
            if not task_id:
                print("❌ 提交切片任务返回缺少 task_id")
                return task_ids
            if status not in ("pending", "processing"):
                print(f"❌ 提交切片任务返回状态异常: {status}")
                return task_ids

            task_ids.append(task_id)
            print(f"✅ 切片任务提交成功: task_id={task_id}")

        except Exception as e:
            print(f"❌ 提交切片任务异常: {e}")
            return task_ids

    return task_ids


def test_query_clip_tasks(task_ids: list[str]) -> bool:
    separator("3. 查询切片任务  GET /api/video-analyze/clip/tasks/{task_id}")
    if not task_ids:
        print("❌ 没有可查询的 task_id")
        return False

    print(f"任务数量: {len(task_ids)}")
    print(f"轮询间隔: {POLL_INTERVAL}s，最多等待: {POLL_MAX_WAIT}s")

    task_status: dict[str, str] = {task_id: "pending" for task_id in task_ids}
    start = time.time()
    while time.time() - start < POLL_MAX_WAIT:
        elapsed = time.time() - start
        completed = 0
        failed = 0

        for task_id in task_ids:
            if task_status.get(task_id) in ("completed", "failed"):
                if task_status[task_id] == "completed":
                    completed += 1
                else:
                    failed += 1
                continue

            url = f"{BASE_URL}/api/video-analyze/clip/tasks/{task_id}"
            try:
                resp = requests.get(url, timeout=POLL_TIMEOUT)
                data = resp.json()
                task_data = data.get("data") or {}
                status = task_data.get("status")
                task_status[task_id] = status

                print(f"\n[{elapsed:5.1f}s] task_id={task_id}  http={resp.status_code}  task_status={status}")
                pretty(data)

                if resp.status_code != 200 or data.get("code") != 200:
                    print(f"❌ 查询切片任务失败: http={resp.status_code} code={data.get('code')}")
                    return False

                if task_data.get("task_type") != "clip":
                    print(f"❌ task_type 异常: {task_data.get('task_type')}")
                    return False

                if status == "completed":
                    ok, message = validate_clip_result(task_data.get("result"))
                    if not ok:
                        print(f"❌ 切片任务结果结构错误: {message}")
                        return False
                    completed += 1
                elif status == "failed":
                    failed += 1
                    print(f"❌ 切片任务失败: {task_data.get('error')}")
                elif status not in ("pending", "processing"):
                    print(f"❌ 未知任务状态: {status}")
                    return False

            except requests.exceptions.Timeout:
                print(f"⚠️  查询超时，继续轮询: task_id={task_id}")
            except Exception as e:
                print(f"❌ 查询切片任务异常: {e}")
                return False

        print(f"\n[{elapsed:5.1f}s] completed={completed}/{len(task_ids)} failed={failed}/{len(task_ids)}")
        if completed + failed == len(task_ids):
            if failed:
                print("❌ 切片任务查询失败，存在 failed 任务")
                return False
            print("✅ 切片任务查询成功，全部任务已完成")
            return True

        time.sleep(POLL_INTERVAL)

    unfinished = [task_id for task_id, status in task_status.items() if status not in ("completed", "failed")]
    print(f"❌ 轮询超时（等待 {POLL_MAX_WAIT}s），未完成任务数: {len(unfinished)}")
    for task_id in unfinished:
        print(f"  - {task_id}: {task_status.get(task_id)}")
    return False


def main() -> None:
    print("=" * 60)
    print("  Video Analyze Clip API 测试")
    print(f"  Base URL: {BASE_URL}")
    print(f"  Video:    {VIDEO_URL}")
    print(f"  Prompt:   {CLIP_PROMPT}")
    print(f"  Async:    {ASYNC_TASK_COUNT} 个切片任务，每 {POLL_INTERVAL}s 轮询")
    print("=" * 60)

    results: dict[str, bool] = {}

    results["同步切片分析"] = test_sync_clip()

    task_ids = test_submit_clip_tasks()
    results["提交切片任务"] = len(task_ids) == ASYNC_TASK_COUNT
    results["查询切片任务"] = test_query_clip_tasks(task_ids) if task_ids else False

    separator("测试结果汇总")
    for name, passed in results.items():
        icon = "✅" if passed else "❌"
        print(f"  {icon}  {name}")

    total = len(results)
    passed = sum(1 for v in results.values() if v)
    print(f"\n  总计: {passed}/{total} 通过")

    sys.exit(0 if passed == total else 1)


if __name__ == "__main__":
    main()
